conf/report/report-template.html

```diff
@@ -42,183 +43,363 @@
+              else if(operator === 'is not'){
+                if(theadname !== tool){
+                  var index = tool_array.indexOf(column.index());
+                  tool_array.splice(index, 1);
+                  selected_tools.push(column.index());
                 }
               }
             }
           });
-          table.columns(array).visible(false);
+          table.columns(tool_array).visible(false);
         }
       }
     }
 
-    function entryChanged(mode, id, parent) {
-      filter_types[mode].value_field_factory(id, parent)
-    }
-
-    function arrayMove(arr, fromIndex, toIndex) {
-      var element = arr[fromIndex];
-      arr.splice(fromIndex, 1);
-      arr.splice(toIndex, 0, element);
-      return arr;
+    function entryChanged(mode, id) {
+      $(`li[data-uid=${id}] span`).remove();
+      filter_types[mode].value_field_factory(id);
     }
 
     function applyFilter() {
+      table.columns(col_array).every(function() {
+        $.fn.dataTable.ext.search.pop();
+      });
       table.columns(col_array).search("").draw();
       table.columns(col_array).visible(true);
       $.fn.dataTable.ext.search.pop();
+      table.draw();
       $('.filter-remove').show();
       var applied_filters = [];
-      var container = $('.filter-entries li select option:selected').each(function() {
-        applied_filters.push($(this).val());
+      $('.filter-entries').find('li').each(function(){
+        var object = {};
+        object['filter'] = $(this).find('.filter-entry-type').val();
+        if (object['filter'] === 'tool')
+          object['priority'] = 1
+        else
+          object['priority'] = 0
+        $(this).find('span').each(function(index){
+          var span = {};
+          span['relation'] = $(this).find('.filter-condition').val();
+          span['operator'] = $(this).find('.filter-entry-operator').val();
+          span['value'] = $(this).find('.filter-entry-value').val();
+          object[index] = span;
+        });
+        applied_filters.push(object);
       });
-
-      if(applied_filters.includes("tool")){
-        tool_selected = true;
-        tool_index = applied_filters.indexOf("tool");
-        applied_filters = arrayMove(applied_filters, tool_index, 0);
-        applied_filters = arrayMove(applied_filters, tool_index+1, 1);
-        applied_filters = arrayMove(applied_filters, tool_index+2, 2);
+      applied_filters.sort((a, b) => (a.priority < b.priority) ? 1 : -1);
+      for(const key in applied_filters){
+        if (applied_filters[key]['filter'] === 'tool')
+          tool_relation = applied_filters[key][0]['relation'];
+        if (applied_filters[key]['filter'] !== 'tool' && (Object.keys(applied_filters[key]).length-2 === 2))
+          filter_types[applied_filters[key]['filter']].apply(applied_filters[key][0]['operator'], applied_filters[key][0]['value'], applied_filters[key][1]['operator'], applied_filters[key][1]['value'], applied_filters[key][0]['relation']);
```

Comment:
I thought that there wouldn't be more than 2 operators/values in the case of coverage and type and implemented it this way. Please let me know if there is a better way to do this.

Reply:
Didn't check how the whole code works, but generally:
* Limit visible/analyzed columns to all columns from the tool filter (relation doesn't matter here)
* For each row in every visible column apply a coverage/type filter. Store the results in a 2d array.
* Apply the tool filter (relations) to every row in the 2d array generated above. Show only rows that pass this filter.

As for variable length filters - create a callback function that can be passed to `Array.reduce()` function. This callback would take accumulated value and a next operator/value pair, and return `accumulator (operator) next_value`. So basically something you already have, but run it in a loop and assign result of each step to "first value" of the next step.